ddl/placement/bundle.go

```diff
@@ -49,6 +49,224 @@ func NewBundle(id int64) *Bundle {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: FollowerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, FollowerRules...)
+	}
+
+	if learnerCount > 0 {
+		LearnerRules, err := NewRules(Learner, learnerCount, learnerConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid LearnerConstraints", err)
+		}
+		for _, rule := range LearnerRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: LearnerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, LearnerRules...)
+	}
+
+	return &Bundle{Rules: Rules}, nil
+}
+
+// NewBundleFromSugarOptions will transform syntax sugar options into the bundle.
+func NewBundleFromSugarOptions(primaryRegion string, regions []string, followers uint64, schedule string) (*Bundle, error) {
+	if followers == 0 {
+		followers = 2
+	}
+	var constraints Constraints
+	var err error
+
+	if len(primaryRegion) == 0 {
+		return nil, fmt.Errorf("%w: empty primary region", ErrInvalidPlacemeOptions)
+	}
+
+	Rules := []*Rule{}
+	switch strings.ToLower(schedule) {
+	case "", "even":
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid PrimaryRegion '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+	case "majority_in_primary":
+		// number of quorum members, without the leader
+		primary := (followers + 1) / 2
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid 'PrimaryRegion', '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+		Rules = append(Rules, NewRule(Follower, primary, constraints))
+		// even split the remaining followers
+		followers = followers - primary
+	default:
+		return nil, fmt.Errorf("%w: unsupported schedule %s", ErrInvalidPlacemeOptions, schedule)
+	}
+
+	if uint64(len(regions)) > followers {
+		return nil, fmt.Errorf("%w: remain %d region to schedule, only %d follower left", ErrInvalidPlacemeOptions, uint64(len(regions)), followers)
+	}
+
+	if len(regions) == 0 {
+		constraints, err := NewConstraints(nil)
+		if err != nil {
+			return nil, err
+		}
+		Rules = append(Rules, NewRule(Follower, followers, constraints))
+	} else {
+		count := followers / uint64(len(regions))
+		rem := followers - count*uint64(len(regions))
+		for i, region := range regions {
+			constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", region)})
+			if err != nil {
+				return nil, fmt.Errorf("%w: invalid region of 'Regions', '%s'", err, region)
+			}
+			if i == 0 && rem > 0 {
+				Rules = append(Rules, NewRule(Follower, count+rem, constraints))
+			} else {
+				Rules = append(Rules, NewRule(Follower, count, constraints))
+			}
+		}
```

Comment:
I'm trying to understand this codes. If user write `PRIMARY_REGION="r1" REGIONS="r2,r3,r4" followers=6` , the rules for followers will be

```
[
  {"role": "follower", "count": 3, "label_constraints": [{"key": "region", "op": "In", values: ["r2"]}]}
  {"role": "follower", "count": 1, "label_constraints": [{"key": "region", "op": "In", values: ["r3"]}]}
  {"role": "follower", "count": 1, "label_constraints": [{"key": "region", "op": "In", values: ["r4"]}]}
]
```

Maybe below is better:

```
[
  {"role": "follower", "count": 2, "label_constraints": [{"key": "region", "op": "In", values: ["r2"]}]}
  {"role": "follower", "count": 2, "label_constraints": [{"key": "region", "op": "In", values: ["r3"]}]}
  {"role": "follower", "count": 1, "label_constraints": [{"key": "region", "op": "In", values: ["r4"]}]}
]
```

Though replica 8 rarely happens in reality but I think it's more even in algorithm level

Reply:
But I will choose a better split algorithm that can balance better for rare cases.

Reply:
Addressed.